wgpu/src/backend/webgpu.rs

```diff
@@ -1417,83 +1523,176 @@ impl crate::context::Context for ContextWebGpu {
         desc: crate::ShaderModuleDescriptor<'_>,
         _shader_bound_checks: wgt::ShaderBoundChecks,
     ) -> (Self::ShaderModuleId, Self::ShaderModuleData) {
-        let mut descriptor: webgpu_sys::GpuShaderModuleDescriptor = match desc.source {
-            #[cfg(feature = "spirv")]
-            crate::ShaderSource::SpirV(ref spv) => {
-                use naga::{back, front, valid};
-
-                let options = naga::front::spv::Options {
-                    adjust_coordinate_space: false,
-                    strict_capabilities: true,
-                    block_ctx_dump_prefix: None,
-                };
-                let spv_parser = front::spv::Frontend::new(spv.iter().cloned(), &options);
-                let spv_module = spv_parser.parse().unwrap();
-
-                let mut validator = valid::Validator::new(
-                    valid::ValidationFlags::all(),
-                    valid::Capabilities::all(),
-                );
-                let spv_module_info = validator.validate(&spv_module).unwrap();
-
-                let writer_flags = naga::back::wgsl::WriterFlags::empty();
-                let wgsl_text =
-                    back::wgsl::write_string(&spv_module, &spv_module_info, writer_flags).unwrap();
-                webgpu_sys::GpuShaderModuleDescriptor::new(wgsl_text.as_str())
-            }
-            #[cfg(feature = "glsl")]
-            crate::ShaderSource::Glsl {
-                ref shader,
-                stage,
-                ref defines,
-            } => {
-                use naga::{back, front, valid};
-
-                // Parse the given shader code and store its representation.
-                let options = front::glsl::Options {
+        fn create_shader_module(
+            desc: &crate::ShaderModuleDescriptor<'_>,
+        ) -> Result<
+            (
+                webgpu_sys::GpuShaderModuleDescriptor,
+                WebShaderCompilationInfo,
+            ),
+            crate::CompilationInfo,
+        > {
+            match desc.source {
+                #[cfg(feature = "spirv")]
+                crate::ShaderSource::SpirV(ref spv) => {
+                    use naga::front;
+
+                    let options = naga::front::spv::Options {
+                        adjust_coordinate_space: false,
+                        strict_capabilities: true,
+                        block_ctx_dump_prefix: None,
+                    };
+                    let spv_parser = front::spv::Frontend::new(spv.iter().cloned(), &options);
+                    let spv_module = spv_parser
+                        .parse()
+                        .map_err(|inner| wgc::pipeline::ShaderError {
+                            source: String::new(),
+                            label: desc.label.map(|s| s.to_string()),
+                            inner: Box::new(inner),
+                        })
```

Comment:
since you asked about less code dupl - this one could have a mapping utility function. A `From`/(`Into`) would be nice, but won't work unfortunately since each type is defined elsewhere :/